src/coreComponents/physicsSolvers/fluidFlow/GeochemicalModel.cpp

```diff
@@ -0,0 +1,819 @@
+    arrayView2d<real64 const> const & dConc         = m_deltaConcentration[er][esr];    
+    arrayView2d<real64 const> const & totalConc     = m_totalConcentration[er][esr];
+
+    arrayView2d<real64 const> const & dependentConc     = m_dependentConc[er][esr][m_reactiveFluidIndex];
+    arrayView3d<real64 const> const & dDependentConc_dConc     = m_dDependentConc_dConc[er][esr][m_reactiveFluidIndex];         
+
+    forAll<serialPolicy>( subRegion.size(), [=] ( localIndex const ei )
+    {
+      if (elemGhostRank[ei] < 0)
+      {
+        stackArray1d<globalIndex, ReactiveFluidBase::MAX_NUM_SPECIES> localAccumDOF( m_numDofPerCell );
+        stackArray1d<real64, ReactiveFluidBase::MAX_NUM_SPECIES> localAccum( m_numDofPerCell );
+        stackArray2d<real64, ReactiveFluidBase::MAX_NUM_SPECIES * ReactiveFluidBase::MAX_NUM_SPECIES> localAccumJacobian( m_numDofPerCell, m_numDofPerCell );
+        
+        globalIndex const elemDOF = dofNumber[ei];
+        
+        for (localIndex idof = 0; idof < m_numDofPerCell; ++idof)
+        {
+          localAccumDOF[idof] = elemDOF + idof;
+        }
+        
+//        localAccumJacobian = 0.0;
+
+        for (localIndex ic = 0; ic < m_numBasisSpecies; ++ic)
+          {
+
+            real64 concBasis = pow(10.0, conc[ei][ic]+dConc[ei][ic]);
+
+            localAccum[ic] = concBasis - totalConc[ei][ic];
+
+            localAccumJacobian[ic][ic] = log(10.0) * concBasis;
+
+            if(isHplus[ic])
+              {
+                localAccum[ic] = 0.0;
+                localAccumJacobian[ic][ic] = 1.0;
+                continue;
+              }
+
+            for (localIndex id = 0; id < m_numDependentSpecies; ++id)
+              {
+                real64 concDependent = pow(10.0, dependentConc[ei][id]);
+                localAccum[ic] -= stochMatrix[ic][id] * concDependent;
+
+                for (localIndex idc = 0; idc < m_numBasisSpecies; ++idc)
+                  {
+
+                    localAccumJacobian[ic][idc] -= stochMatrix[ic][id] * log(10.0) * concDependent * dDependentConc_dConc[ei][id][idc];
+
+                  }
+              }
+          }
+
+
+        // add contribution to global residual and jacobian
+//        rhs->add( localAccumDOF.data(),
+//                  localAccum.data(),
+//                  m_numDofPerCell );
+//
+//        matrix->add( localAccumDOF.data(),
+//                     localAccumDOF.data(),
+//                     localAccumJacobian.data(),
+//                     m_numDofPerCell, m_numDofPerCell );
+
+        for( localIndex i = 0; i < m_numDofPerCell; ++i )
+        {
+          localRhs[localAccumDOF[i]] += localAccum[i];
+          localMatrix.addToRow< serialAtomic >( localAccumDOF[i],
+                                                localAccumDOF.data(),
+                                                localAccumJacobian[i],
+                                                m_numDofPerCell );
```

Comment:
@klevzoff Does this look correct?

Reply:
Yes, assuming this solver is coupled with others iteratively. Otherwise needs the usual local row offsetting:
```suggestion
        for( localIndex i = 0; i < m_numDofPerCell; ++i )
        {
          localIndex const localRow = localAccumDOF[i] - rankOffset;
          localRhs[localRow] += localAccum[i];
          localMatrix.addToRow< serialAtomic >( localRow,
                                                localAccumDOF.data(),
                                                localAccumJacobian[i],
                                                m_numDofPerCell );
```